core/src/main/java/software/amazon/awssdk/core/async/ByteArrayAsyncRequestBody.java

```diff
@@ -43,20 +45,42 @@ public long contentLength() {
     }
 
     @Override
-    public void subscribe(Subscriber<? super ByteBuffer> subscriber) {
-        subscriber.onSubscribe(
-                new Subscription() {
-                    @Override
-                    public void request(long n) {
-                        if (n > 0) {
-                            subscriber.onNext(ByteBuffer.wrap(bytes));
-                            subscriber.onComplete();
+    public void subscribe(Subscriber<? super ByteBuffer> s) {
+        // As per rule 1.9 we must throw NullPointerException if the subscriber parameter is null
+        if (s == null) {
+            throw new NullPointerException("Subscription MUST NOT be null.");
+        }
+
+        // As per 2.13, this method must return normally (i.e. not throw).
+        try {
+            s.onSubscribe(
+                    new Subscription() {
+                        boolean done = false;
+
+                        @Override
+                        public void request(long n) {
+                            if (n > 0) {
+                                if (!done) {
+                                    s.onNext(ByteBuffer.wrap(bytes));
+                                    done = true;
+                                    s.onComplete();
+                                }
+                            } else {
+                                s.onError(new IllegalArgumentException("§3.9: non-positive requests are not allowed!"));
+                            }
                         }
-                    }
 
-                    @Override
-                    public void cancel() {
+                        @Override
+                        public void cancel() {
+                        }
                     }
-                });
+            );
+        } catch (Throwable ex) {
+            new IllegalStateException(s + " violated the Reactive Streams rule 2.13 " +
```

Comment:
do we want to log `Subscriber` here? the implementations might not override `toString`

Reply:
I think it's okay, it's just a diagnostic. Correct implementations should return normally from `onSubscribe` so it should be it should be rare that we hit this